## Creating the machine learning algorithm

### Import the relevant libraries

In [1]:
import numpy as np
import tensorflow as tf

### Data

In [4]:
npz = np.load('Audiobooks_data_train.npz')
train_inputs = npz['inputs'].astype(float)
train_targets = npz['targets'].astype(int)

npz = np.load('Audiobooks_data_validation.npz')
validation_inputs, validation_targets = npz['inputs'].astype(float), npz['targets'].astype(int)

npz = np.load('Audiobooks_data_test.npz')
test_inputs, test_targets = npz['inputs'].astype(float), npz['targets'].astype(int)

### Model
Outline, optimizers, loss, early stopping and training

In [5]:
# input_size = 10
output_size = 2
hidden_layer_size = 50
    
model = tf.keras.Sequential([
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'), # 1st hidden layer
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'), # 2nd hidden layer
                            tf.keras.layers.Dense(output_size, activation='softmax') # output layer
                            ])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

batch_size = 100

max_epochs = 100

early_stopping = tf.keras.callbacks.EarlyStopping(patience=2)

model.fit(train_inputs, 
          train_targets, 
          batch_size=batch_size, 
          epochs=max_epochs, 
          callbacks=[early_stopping], 
          validation_data=(validation_inputs, validation_targets), 
          verbose = 2 
          )  

Epoch 1/100
36/36 - 1s - loss: 0.4974 - accuracy: 0.8357 - val_loss: 0.4180 - val_accuracy: 0.8434 - 821ms/epoch - 23ms/step
Epoch 2/100
36/36 - 0s - loss: 0.3433 - accuracy: 0.8810 - val_loss: 0.3654 - val_accuracy: 0.8546 - 116ms/epoch - 3ms/step
Epoch 3/100
36/36 - 0s - loss: 0.3051 - accuracy: 0.8891 - val_loss: 0.3413 - val_accuracy: 0.8658 - 112ms/epoch - 3ms/step
Epoch 4/100
36/36 - 0s - loss: 0.2863 - accuracy: 0.8977 - val_loss: 0.3295 - val_accuracy: 0.8702 - 110ms/epoch - 3ms/step
Epoch 5/100
36/36 - 0s - loss: 0.2740 - accuracy: 0.8980 - val_loss: 0.3107 - val_accuracy: 0.8792 - 109ms/epoch - 3ms/step
Epoch 6/100
36/36 - 0s - loss: 0.2645 - accuracy: 0.9019 - val_loss: 0.3078 - val_accuracy: 0.8814 - 112ms/epoch - 3ms/step
Epoch 7/100
36/36 - 0s - loss: 0.2587 - accuracy: 0.9053 - val_loss: 0.3028 - val_accuracy: 0.8837 - 106ms/epoch - 3ms/step
Epoch 8/100
36/36 - 0s - loss: 0.2564 - accuracy: 0.9047 - val_loss: 0.2976 - val_accuracy: 0.8859 - 100ms/epoch - 3ms/step
Epoch 9

## 89.26% validation accuracy is very good!

## Testing the Model

In [6]:
test_loss, test_accuracy = model.evaluate(test_inputs, test_targets)

14/14 [==============================] - 0s 2ms/step - loss: 0.2410 - accuracy: 0.9062


In [7]:
print('\nTest loss: {0:.2f}. Test accuracy: {1:.2f}%'.format(test_loss, test_accuracy*100.))


Test loss: 0.24. Test accuracy: 90.62%


## Test accuracy is 90.62%. It is slightly more than the validation accuracy but its a really small difference so nothing to worry.

## Obtaining the probability for a customer to convert

In [9]:
model.predict(test_inputs).round(2)

14/14 [==============================] - 0s 2ms/step


array([[0.  , 1.  ],
       [1.  , 0.  ],
       [1.  , 0.  ],
       [0.19, 0.81],
       [0.92, 0.08],
       [0.72, 0.28],
       [0.92, 0.08],
       [0.11, 0.89],
       [0.  , 1.  ],
       [0.31, 0.69],
       [0.95, 0.05],
       [0.84, 0.16],
       [0.73, 0.27],
       [0.  , 1.  ],
       [0.98, 0.02],
       [0.91, 0.09],
       [0.91, 0.09],
       [1.  , 0.  ],
       [0.08, 0.92],
       [0.16, 0.84],
       [1.  , 0.  ],
       [0.  , 1.  ],
       [0.88, 0.12],
       [1.  , 0.  ],
       [0.  , 1.  ],
       [0.75, 0.25],
       [0.12, 0.88],
       [0.  , 1.  ],
       [0.88, 0.12],
       [0.  , 1.  ],
       [1.  , 0.  ],
       [0.22, 0.78],
       [1.  , 0.  ],
       [0.65, 0.35],
       [0.94, 0.06],
       [0.06, 0.94],
       [0.  , 1.  ],
       [0.13, 0.87],
       [0.82, 0.18],
       [0.04, 0.96],
       [0.26, 0.74],
       [0.93, 0.07],
       [0.91, 0.09],
       [0.23, 0.77],
       [0.88, 0.12],
       [0.47, 0.53],
       [0.93, 0.07],
       [0.03,

- first column is probability of classified as 0 and second column is probability of classified as 1. 

In [11]:
# Alternatively, we can get only the second column, that is if the customer will convert again.
model.predict(test_inputs)[:,1].round(2)

14/14 [==============================] - 0s 2ms/step


array([1.  , 0.  , 0.  , 0.81, 0.08, 0.28, 0.08, 0.89, 1.  , 0.69, 0.05,
       0.16, 0.27, 1.  , 0.02, 0.09, 0.09, 0.  , 0.92, 0.84, 0.  , 1.  ,
       0.12, 0.  , 1.  , 0.25, 0.88, 1.  , 0.12, 1.  , 0.  , 0.78, 0.  ,
       0.35, 0.06, 0.94, 1.  , 0.87, 0.18, 0.96, 0.74, 0.07, 0.09, 0.77,
       0.12, 0.53, 0.07, 0.97, 0.11, 0.78, 0.13, 0.  , 0.2 , 0.  , 0.88,
       0.11, 0.  , 1.  , 0.09, 0.8 , 0.08, 0.99, 0.77, 0.79, 1.  , 0.78,
       0.04, 0.05, 0.25, 0.96, 0.1 , 0.  , 0.4 , 1.  , 0.94, 0.84, 1.  ,
       1.  , 0.  , 1.  , 1.  , 0.78, 0.08, 1.  , 0.96, 0.91, 0.04, 0.  ,
       0.98, 0.97, 0.87, 0.15, 0.88, 0.  , 0.26, 0.95, 0.99, 0.95, 0.07,
       0.07, 0.79, 0.  , 0.07, 0.01, 0.98, 0.89, 0.  , 0.17, 0.97, 0.99,
       0.87, 0.91, 0.02, 0.  , 0.03, 0.78, 0.08, 0.  , 0.  , 0.29, 1.  ,
       0.1 , 0.82, 0.05, 0.21, 0.  , 0.13, 0.84, 0.08, 0.95, 0.81, 0.  ,
       0.77, 0.23, 0.83, 0.09, 1.  , 1.  , 0.28, 0.77, 0.01, 1.  , 1.  ,
       0.94, 0.09, 0.51, 0.65, 0.74, 0.05, 1.  , 0.

In [12]:
model.predict(test_inputs)[:,1].round(0)

14/14 [==============================] - 0s 2ms/step


array([1., 0., 0., 1., 0., 0., 0., 1., 1., 1., 0., 0., 0., 1., 0., 0., 0.,
       0., 1., 1., 0., 1., 0., 0., 1., 0., 1., 1., 0., 1., 0., 1., 0., 0.,
       0., 1., 1., 1., 0., 1., 1., 0., 0., 1., 0., 1., 0., 1., 0., 1., 0.,
       0., 0., 0., 1., 0., 0., 1., 0., 1., 0., 1., 1., 1., 1., 1., 0., 0.,
       0., 1., 0., 0., 0., 1., 1., 1., 1., 1., 0., 1., 1., 1., 0., 1., 1.,
       1., 0., 0., 1., 1., 1., 0., 1., 0., 0., 1., 1., 1., 0., 0., 1., 0.,
       0., 0., 1., 1., 0., 0., 1., 1., 1., 1., 0., 0., 0., 1., 0., 0., 0.,
       0., 1., 0., 1., 0., 0., 0., 0., 1., 0., 1., 1., 0., 1., 0., 1., 0.,
       1., 1., 0., 1., 0., 1., 1., 1., 0., 1., 1., 1., 0., 1., 0., 0., 0.,
       0., 1., 0., 0., 1., 0., 0., 0., 0., 1., 1., 1., 0., 1., 0., 1., 1.,
       1., 1., 1., 1., 1., 0., 0., 0., 0., 1., 0., 1., 1., 1., 0., 0., 0.,
       1., 1., 0., 1., 0., 0., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 1.,
       1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 0., 1., 0., 1., 0., 1., 0.,
       0., 1., 1., 0., 0.

In [13]:
# A much better approach here would be to use argmax 
# This method is great for multi-class problems as it is independent of the number of classes
np.argmax(model.predict(test_inputs),axis=1)

14/14 [==============================] - 0s 1ms/step


array([1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0,
       0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0,
       1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0,

## Saving the model

In [14]:
model.save('audiobooks_model.h5') 